src/apps/intel/intel10g.lua

```diff
@@ -224,11 +248,102 @@ end
 
 --- See datasheet section 7.1 "Inline Functions -- Transmit Functionality."
 
+--- "Keeps offload context"
+--- if the packet requests checksums to be performed,
+--- this function makes sure the current offload context
+--- correctly describes the packet type.  if needed
+--- it sends a new context descriptor.
+local function keep_offload_ctx(self, p)
+   if band(p.flags, C.PACKET_NEEDS_CSUM) == 0 then
+      self.prev_ctx_desc_a = 0ULL
+      self.prev_ctx_desc_b = 0ULL
+      self.offloadflags = 0ULL
+      return
+   end
+   local b = p.data
+   local ctx_desc_a = 0ULL
+   local ctx_desc_b = 0x0000000fe0200000ULL -- DTYP=2, DEXT=1, BCNTLEN=0x3F
+   local iplen, maclen, ctx_l4t = 0ULL, 14ULL, 3ULL
+   local ethtype, proto = ffi.cast('uint16_t *', b+ETHERTYPE_OFFSET)[0], 0ULL
+   local offloadflags = 0ULL
+
+   if ethtype == ETHERTYPE_IPV4 then
+      ctx_desc_b = bor(ctx_desc_b, 0x0400ULL)     -- TUCMD.IPV4
+      iplen = lshift(band(b[IPV4_IHL_OFFSET], 0xF), 2)
+      offloadflags = bor(offloadflags, lshift(1ULL, 40))      -- IXSM
+      proto = b[IPV4_PROTOCOL_OFFSET]
+      b[IPV4_CSUM_OFFSET] = 0
+      b[IPV4_CSUM_OFFSET+1] = 0
+
+   elseif ethtype == ETHERTYPE_IPV6 then
+      iplen = 40ULL
+      proto = b[IPV6_NEXT_HEADER_OFFSET]
+   end
+
+   if proto == IP_TCP then
+      ctx_l4t = 0x01ULL
+      offloadflags = bor(offloadflags, lshift(1ULL, 41))      -- TXSM
+      b[maclen + iplen + TCP_CSUM_OFFSET] = 0
+      b[maclen + iplen + TCP_CSUM_OFFSET+1] = 0
+   elseif proto == IP_UDP then
+      ctx_l4t = 0x00ULL
+      offloadflags = bor(offloadflags, lshift(1ULL, 41))      -- TXSM
+      b[maclen + iplen + UDP_CSUM_OFFSET] = 0
+      b[maclen + iplen + UDP_CSUM_OFFSET+1] = 0
+   elseif proto == IP_SCTP then
+      ctx_l4t = 0x02ULL
+      offloadflags = bor(offloadflags, lshift(1ULL, 41))      -- TXSM
+      b[maclen + iplen + SCTP_CSUM_OFFSET] = 0
+      b[maclen + iplen + SCTP_CSUM_OFFSET+1] = 0
+   end
+
+   ctx_desc_a = bor(ctx_desc_a, lshift(band(maclen, 0x7F), 9), band(iplen, 0x1FF))
+   ctx_desc_b = bor(ctx_desc_b, lshift(ctx_l4t, 11))
+   if ctx_desc_a ~= self.prev_ctx_desc_a or ctx_desc_b ~= self.prev_ctx_desc_b then
+      self.txdesc[self.tdt].address = ctx_desc_a
+      self.txdesc[self.tdt].options = ctx_desc_b
+      self.txpackets[self.tdt] = nil
+      self.tdt = band(self.tdt + 1, num_descriptors - 1)
+      self.prev_ctx_desc_a = ctx_desc_a
+      self.prev_ctx_desc_b = ctx_desc_b
+      self.offloadflags = offloadflags
+   end
+end
+
+-- local txdesc_flags = 0x0000000022300000ULL                      -- DEXT=1, IFCS=1, DTYP=0011b
```

Comment:
right, i'm still cleaning it up.  during the after-straightline rebase there were some conflicts that confused me, so i commented out most to later compare with the previous history.